addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,441 @@
+
+	// make sure the TKR object exists
+	tkrName, err := util.GetClusterLabel(clusterLabels, constants.TKRLabelClassyClusters)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	tkr, err := util.GetTKRByName(r.ctx, r.Client, tkrName)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "unable to fetch TKR object '%s'", tkrName)
+	}
+	// if tkr is not found, should not requeue for the reconciliation
+	if tkr == nil {
+		log.Info("TKR object not found", "name", tkrName)
+		return ctrl.Result{}, nil
+	}
+
+	if cluster.Status.Phase != string(clusterapiv1beta1.ClusterPhaseProvisioned) {
+		log.Info(fmt.Sprintf("cluster %s/%s does not have status phase %s", cluster.Namespace, cluster.Name, clusterapiv1beta1.ClusterPhaseProvisioned))
+		return ctrl.Result{}, nil
+	}
+
+	// determine cluster role
+	_, isManagementCluster := clusterLabels[tkrconstants.ManagementClusterRoleLabel]
+
+	if isManagementCluster {
+		// the cluster is a management cluster
+		clusterRole = clusterRoleManagement
+		clusterClient = r.Client
+	} else {
+		// the cluster is a remote workload cluster
+		clusterRole = clusterRoleWorkload
+		remoteClient, err := r.tracker.GetClient(r.ctx, clusterapiutil.ObjectKey(cluster))
+		if err != nil {
+			return ctrl.Result{RequeueAfter: time.Second * 10}, errors.Wrap(err, "error getting remote cluster's client")
+		}
+		clusterClient = remoteClient
+		log.Info("successfully got remoteClient")
+		// set watch if not already set. If the watch already exists, it doesn't get re-created
+		if err := r.watchPackageInstalls(cluster, log); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "error watching PackageInstalls on target cluster")
+		}
+		log.Info("finished setting up remote watch for the cluster")
+	}
+
+	if err := r.reconcile(clusterClient, cluster, clusterRole, log); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcile iterates over all core/additional packages and calls reconcileClusterBootstrapStatus() for each.
+// it eventually patches ClusterBootstrapStatus with the condition entries (one for each package) in a single patch operation
+func (r *PackageInstallStatusReconciler) reconcile(clusterClient client.Client, cluster *clusterapiv1beta1.Cluster, clusterRole ClusterRole, log logr.Logger) (retErr error) {
+	clusterObjKey := client.ObjectKeyFromObject(cluster)
+
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	// ClusterBootstrap resource exists in the management cluster, that's why we need to use local client instead of remote client for fetching it
+	if err := r.Client.Get(r.ctx, clusterObjKey, clusterBootstrap); err != nil {
+		return err
+	}
+
+	patchHelper, err := clusterapipatchutil.NewHelper(clusterBootstrap, r.Client)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		log.Info("patching ClusterBootstrapStatus")
+		if err := patchHelper.Patch(r.ctx, clusterBootstrap); err != nil {
+			retErr = errors.Wrap(err, "error patching ClusterBootstrapStatus")
```

Comment:
(looks like there is currently no `return nil, err`, but it's more future proof to append the error instead of overwriting it (kerrors.NewAggreate will handle the case if reterr is nil)

(`kerrors "k8s.io/apimachinery/pkg/util/errors"`)